File: 2023/day04/solution.py
from datetime import datetime
from dataclasses import dataclass
from typing import List


@dataclass
class Card:
    id: int
    wins: int


def cards(filename: str) -> List[Card]:
    out = []
    with open(filename) as f:
        for line in f.readlines():
            line = line.strip()
            c, line = line.split(": ")
            winning, have = line.split(" | ")
            num = int(c.split()[1])
            winning = {int(x) for x in winning.split()}
            have = {int(x) for x in have.split()}
            out.append(Card(num, len(have.intersection(winning))))
    return out


def p1(cards: List[Card]) -> int:
    return sum([2**(c.wins-1) for c in cards if c.wins > 0])


def p2(cards: List[Card]) -> int:
    total = len(cards)
    for c in cards:
        total += wins(c, cards)
    return total


def cached(func):
    cache = {}

    def wrapper(card: Card, cards: List[Card]):
        if card.id in cache:
            return cache[card.id]
        res = func(card, cards)
        cache[card.id] = res
        return res

    return wrapper


@cached
def wins(card: Card, cards: List[Card]):
    total = card.wins
    for i in range(card.wins):
        if card.id+i >= len(cards):
            break
        total += wins(cards[card.id+i], cards)
    return total


c = cards("assets/input.txt")
start = datetime.now()
for x in range(10_000):
    p2(c)
print((datetime.now() - start).total_seconds())

File: 2023/day04/solution.go
package main

import (
	"bufio"
	"fmt"
	"math"
	"os"
	"strconv"
	"strings"
	"time"
)

func main() {
	c, err := cards("assets/input.txt")
	if err != nil {
		panic(err)
	}
	fmt.Println(p2(c))
	start := time.Now()
	for i := 0; i < 10000; i++ {
		p2(c)
	}
	fmt.Println(time.Since(start).Seconds())
}

func p1(cards []*Card) int {
	total := 0.0
	for _, c := range cards {
		if c.wins > 0 {
			total += math.Pow(2, float64(c.wins-1))
		}
	}
	return int(total)
}

func p2(cards []*Card) int {
	cache := Cache{
		cache: make([]*int, len(cards)+1),
	}
	total := len(cards)
	for _, card := range cards {
		total += cache.wins(card, cards)
	}
	return total
}

type Card struct {
	id   int
	wins int
}

type Cache struct {
	cache []*int
}

func (c *Cache) wins(card *Card, cards []*Card) int {
	if wins := c.cache[card.id]; wins != nil {
		return *wins
	}
	total := card.wins
	for i := 0; i < card.wins; i++ {
		if card.id+i >= len(cards) {
			break
		}
		total += c.wins(cards[card.id+i], cards)
	}
	c.cache[card.id] = &total
	return total
}

func cards(filename string) ([]*Card, error) {
	f, err := os.Open(filename)
	if err != nil {
		return nil, err
	}
	defer f.Close()
	scanner := bufio.NewScanner(f)
	var out []*Card
	for scanner.Scan() {
		line := scanner.Text()
		card, nums, _ := strings.Cut(line, ": ")
		fields := strings.Fields(card)
		var c Card
		c.id, _ = strconv.Atoi(fields[1])

		winningStr, haveStr, _ := strings.Cut(nums, " | ")

		winning := strToInts(winningStr)
		have := strToInts(haveStr)
		for k := range have {
			if _, ok := winning[k]; ok {
				c.wins++
			}
		}
		out = append(out, &c)
	}
	return out, nil
}

func strToInts(s string) map[int]struct{} {
	f := strings.Fields(s)
	out := make(map[int]struct{}, len(f))
	for _, field := range f {
		k, _ := strconv.Atoi(field)
		out[k] = struct{}{}
	}
	return out
}
